internal/bzlmod/go_deps.bzl

```diff
@@ -276,6 +282,49 @@ _go_repository_config = repository_rule(
     },
 )
 
+def fail_on_version_conflict(version, previous, module_tag, module_name_to_go_dot_mod_label, fail_or_warn):
```

Comment:
"fail_or_warn" makes it hard to figure out what to pass for what behavior.

Instead, how about:
* call this function `check_for_version_conflict`
* call the last argument `conflict_printer` and pass in the `print` or `fail` function at the call site

Reply:
I like it!